### PR TITLE
Raunak/gas griefing fixes

### PR DESCRIPTION
**Pull request by RnkSngh**
PR to fix the gas griefing vuln in dispatcher contract by checking that gasLeft is still > 1/64 of what the low level call took

## Summary by CodeRabbit

- **New Features**
	- Improved gas management in blockchain transactions to ensure sufficient gas is available post-transaction.
	- Introduced error handling for scenarios with insufficient gas during transactions.

- **Tests**
	- Added tests for gas management and behavior under gas-constrained conditions in blockchain interactions.

### CODE REVIEW

File: contracts/core/Dispatcher.sol [hunk 1]
```diff
@@ -555,7 +555,18 @@ contract Dispatcher is OwnableUpgradeable, UUPSUpgradeable, IDispatcher {
...
         // Note: This tx won't revert if the low-level call fails, see
         // https://docs.soliditylang.org/en/latest/cheatsheet.html#members-of-address
+        uint256 gasBeforeCall = gasleft();
         (success, message) = receiver.call(args);
+
+        // x is the amount of gas left right before the low level call. Solidity will allocate x*63/64 to the low level
+        // call
+        // and x/64 for the remaining execution after the low-level call. If this low-level call runs out of gas, then
+        // gasLeft will be equal to x/64 at the start of the remaining execution. so we should check gasBefore < 1/64
+        if (!success && gasleft() < gasBeforeCall / 64) {
```
Comment (dshiell) [inline contracts/core/Dispatcher.sol:565]: <= ?

**Review by dshiell — APPROVED** [review]
lgtm, just wondering about if the gas check should be "<="

File: test/mocks/GasUsingMars.sol [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        // This function is used to test the gas usage of the contract
+        uint256 startingGas = gasleft();
+        uint256 dummyInt = 0;
+
+        // It will use the gasToUse amount of gas
+        while (startingGas - gasleft() < gasToUse) {
+            dummyInt += 1;
+        }
+    }
+}
```
Comment (alfredo-stonk) [inline test/mocks/GasUsingMars.sol:52]: nice way to burn gas 😆

**Review by alfredo-stonk — APPROVED** [review]
Change looks good! 
Need to fix `forge test` though

File: contracts/core/Dispatcher.sol (hunk 1 above)
Comment (RnkSngh) [reply contracts/core/Dispatcher.sol]: Ahh good catch; should be fixed now!

**Review by nicopernas — APPROVED** [review]

**Comment by RnkSngh** [comment]
> Need to fix forge test though
Should be fixed now! Seemed to be a problem with my local os being non-case sensitive but the CI being case sensitive

File: contracts/core/Dispatcher.sol [hunk 3]
```diff
@@ -555,7 +555,18 @@ contract Dispatcher is OwnableUpgradeable, UUPSUpgradeable, IDispatcher {
...
         // Note: This tx won't revert if the low-level call fails, see
         // https://docs.soliditylang.org/en/latest/cheatsheet.html#members-of-address
+        uint256 gasBeforeCall = gasleft();
         (success, message) = receiver.call(args);
+
+        // x is the amount of gas left right before the low level call. Solidity will allocate x*63/64 to the low level
+        // call
+        // and x/64 for the remaining execution after the low-level call. If this low-level call runs out of gas, then
+        // gasLeft will be equal to x/64 at the start of the remaining execution. so we should check gasBefore < 1/64
+        if (!success && gasleft() <= gasBeforeCall / 64) {
```
Comment (reednaa) [inline contracts/core/Dispatcher.sol:565]: While being a very small issue, it should be `gasBeforeCall / 63` not `gasBeforeCall / 64`.

See arguments in our implementation: https://github.com/catalystdao/GeneralisedIncentives/blob/f71a3ba9a50fa1ec3eec8b3409aa98ec625a199b/src/IncentivizedMessageEscrow.sol#L586

**Comment by reednaa** [comment]
Lgtm. Thanks for the fix.

**Pull request merged**